### DISCUSSION TITLE
How are we supposed to know the syntax of python-requires in pyproject.toml?

### DISCUSSION THREAD

**Discussion by metaperl** [pdm-project/pdm#1369] [Q&A]
I got the following error trace:

```
C:\programming\ogs-buddy\bin>pdm add auto-selenium
Adding packages to default dependencies: auto-selenium
  Lock failed
Unable to find a resolution because the following dependencies don't work on all Python versions defined by the
project's `requires-python`: .
  python>=3.6 (from <Candidate auto-selenium 1.0.0 from https://pypi.org/simple/auto-selenium/>)
  python>=3.6 (from <Candidate auto-selenium 1.0.1 from https://pypi.org/simple/auto-selenium/>)
To fix this, you can change the value of `requires-python` in pyproject.toml.
See C:\Users\thequ\AppData\Local\Temp\pdm-lock-qjw3jp5d.log for detailed debug log.
[ResolutionImpossible]: Unable to find a resolution
Add '-v' to see the detailed traceback
```

but then did not know the syntax for `python-requires` in pyproject.toml.

Also [searching at fming](https://pdm.fming.dev/) doesnt help either.

**Comment by pawamoy** [comment]
There's a bit of info here: https://pdm.fming.dev/latest/usage/project/#how-requires-python-controls-the-project

> PDM respects the value of requires-python in the way that it tries to pick package candidates that can work on all python versions that requires-python contains. For example, if requires-python is >=2.7, PDM will try to find the latest version of foo, whose requires-python version range is a superset of >=2.7.
So, make sure you write requires-python properly if you don't want any outdated packages to be locked.

What `python-requires` do you have in `pyproject.toml`?

**Reply by viceice** [reply]
we've `>=3.13` and installed python `3.14.4` globally
https://github.com/renovatebot/docker-build-cache/blob/dfd1f1e081165313957c76bcbd134ac18559959c/pyproject.toml#L5

but pdm is still complaining: `WARNING: Project requires a python version of >=3.13, The virtualenv is being created for you as it cannot be matched to the right version.`

shouldn't it simply use the version from path because it matches?

**Comment by ShreeChaturvedi** [comment]
It's `requires-python` under `[project]` in `pyproject.toml`, not `python-requires`. Drop this in:

```toml
[project]
requires-python = ">=3.8"
```

The `.` in the error is because that field is blank, so pdm tries to lock for every Python. auto-selenium only covers `>=3.6`. Use your real version (3.8, 3.10, whatever you've got). If `[project]` is already there, just add the one line, then `pdm add auto-selenium` again.

Same idea as setuptools' `python_requires`. There's a snippet on the [Python version](https://pdm-project.org/en/latest/reference/pep621/#python-version) page too.
